clippy_lints/src/ptr_offset_with_cast.rs

```diff
@@ -0,0 +1,130 @@
+///
+/// **Known problems:** None
+///
+/// **Example:**
+/// ```rust
+/// let vec = vec![b'a', b'b', b'c'];
+/// let ptr = vec.as_ptr();
+/// let offset = 1_usize;
+///
+/// unsafe { ptr.offset(offset as isize); }
+/// ```
+///
+/// Could be written:
+///
+/// ```rust
+/// let vec = vec![b'a', b'b', b'c'];
+/// let ptr = vec.as_ptr();
+/// let offset = 1_usize;
+///
+/// unsafe { ptr.add(offset); }
+/// ```
+declare_clippy_lint! {
+    pub PTR_OFFSET_WITH_CAST,
+    style,
+    "uneeded pointer offset cast"
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct Pass;
+
+impl lint::LintPass for Pass {
+    fn get_lints(&self) -> lint::LintArray {
+        lint_array!(PTR_OFFSET_WITH_CAST)
+    }
+}
+
+impl<'a, 'tcx> lint::LateLintPass<'a, 'tcx> for Pass {
+    fn check_expr(&mut self, cx: &lint::LateContext<'a, 'tcx>, expr: &'tcx hir::Expr) {
+        // Check if the expressions is a ptr.offset method call
+        let [receiver_expr, arg_expr] = match expr_as_ptr_offset_call(cx, expr) {
+            Some(call_arg) => call_arg,
+            None => return,
+        };
+
+        // Check if the parameter to ptr.offset is a cast from usize to isize
+        let cast_lhs_expr = match expr_as_cast_from_usize(cx, arg_expr) {
+            Some(cast_lhs_expr) => cast_lhs_expr,
+            None => return,
+        };
+
+        utils::span_lint_and_sugg(
+            cx,
+            PTR_OFFSET_WITH_CAST,
+            expr.span,
+            "use of `offset` with a `usize` casted to an `isize`",
+            "try",
+            build_suggestion(cx, receiver_expr, cast_lhs_expr),
+        );
+    }
+}
+
+// If the given expression is a cast from a usize, return the lhs of the cast
+fn expr_as_cast_from_usize<'a, 'tcx>(
+    cx: &lint::LateContext<'a, 'tcx>,
+    expr: &'tcx hir::Expr,
+) -> Option<&'tcx hir::Expr> {
+    if let hir::ExprKind::Cast(ref cast_lhs_expr, _) = expr.node {
+        if is_expr_ty_usize(cx, &cast_lhs_expr) {
+            return Some(cast_lhs_expr);
+        }
+    }
+    None
+}
+
+// If the given expression is a ptr::offset method call, return the receiver and the arg of the
+// method call.
+fn expr_as_ptr_offset_call<'a, 'tcx>(
+    cx: &lint::LateContext<'a, 'tcx>,
+    expr: &'tcx hir::Expr,
+) -> Option<[&'tcx hir::Expr; 2]> {
+    if let hir::ExprKind::MethodCall(ref path_segment, _, ref args) = expr.node {
+        if path_segment.ident.name == "offset" && is_expr_ty_raw_ptr(cx, &args[0]) {
+            return Some([&args[0], &args[1]]);
+        }
+    }
+    None
+}
+
+// Is the type of the expression a usize?
+fn is_expr_ty_usize<'a, 'tcx>(
+    cx: &lint::LateContext<'a, 'tcx>,
+    expr: &hir::Expr,
+) -> bool {
+    cx.tables.expr_ty(expr).sty == ty::TyKind::Uint(ast::UintTy::Usize)
+}
+
+// Is the type of the expression a raw pointer?
+fn is_expr_ty_raw_ptr<'a, 'tcx>(
+    cx: &lint::LateContext<'a, 'tcx>,
+    expr: &hir::Expr,
+) -> bool {
+    if let ty::RawPtr(..) = cx.tables.expr_ty(expr).sty {
```

Comment:
You can use the [`is_unsafe_ptr` method](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyS.html#method.is_unsafe_ptr) instead of destructuring

Reply:
https://github.com/rust-lang-nursery/rust-clippy/pull/3104/commits/f7d1ef90a04ad40148dc0db4ccf739e8178b11b1